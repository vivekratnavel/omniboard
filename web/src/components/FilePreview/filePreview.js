import React, {PureComponent} from 'react';
import {tomorrow} from 'react-syntax-highlighter/styles/prism';
import PropTypes from 'prop-types';
import SyntaxHighlighter, {registerLanguage} from 'react-syntax-highlighter/prism-light';
import json from 'react-syntax-highlighter/languages/prism/json';
import python from 'react-syntax-highlighter/languages/prism/python';
import {Alert} from 'react-bootstrap';
import Iframe from 'react-iframe';
import {ProgressWrapper} from '../Helpers/hoc';
import {FILE_PREVIEW_LIMIT} from '../SourceFilesView/sourceFilesView';
import {getFileExtension} from '../Helpers/utils';
import {imageExtensions} from './imageExtensions';

registerLanguage('json', json);
registerLanguage('python', python);

const getLanguageFromFileName = fileName => {
  const languageMapping = {
    py: 'python',
    json: 'json'
  };
  const extension = fileName.split('.').splice(-1)[0];
  return languageMapping[extension] || '';
};

class FilePreview extends PureComponent {
  static propTypes = {
    fileName: PropTypes.string.isRequired,
    sourceFiles: PropTypes.object.isRequired,
    fileId: PropTypes.string.isRequired,
    isLoading: PropTypes.bool.isRequired,
    errorMessage: PropTypes.string,
    dbInfo: PropTypes.shape({
      path: PropTypes.string,
      key: PropTypes.string
    }).isRequired
  };

  render() {
    const {fileName, sourceFiles, fileId, errorMessage, isLoading, dbInfo} = this.props;
    const extension = getFileExtension(fileName);
    const fileInfo = sourceFiles[fileId];
    const isFileTooLarge = fileInfo && fileInfo.fileLength && fileInfo.fileLength > FILE_PREVIEW_LIMIT;
    const warningMessage = `File is too large to be previewed. Only files less than ${FILE_PREVIEW_LIMIT / 1024 / 1024}MB can be previewed.`;
    const formattedFilePreview = () => {
      if (isFileTooLarge) {
        return (<Alert bsStyle='warning'>{warningMessage}</Alert>);
      }

      const pathPrefix = dbInfo.path ? `${dbInfo.path}/` : '';
      if (imageExtensions.includes(extension) && fileId) {
        const imgSource = `${pathPrefix}api/v1/files/download/${fileId}/${fileName}`;
        return (<img src={imgSource} alt='image'/>);
      }

      // Render html in iframe
      if (extension === 'html') {
        return (
          <Iframe
            url={`${pathPrefix}api/v1/files/preview/${fileId}/${fileName}`}
            width='100%'
            height='1000px'
          />
        );
      }

      if (fileInfo && fileInfo.data) {
        return (<SyntaxHighlighter language={getLanguageFromFileName(fileName)} style={tomorrow}>{fileInfo.data}</SyntaxHighlighter>);
      }

      return null;
    };

    return (
      <div>
        <ProgressWrapper loading={isLoading}>
          {
            errorMessage ?
              <Alert bsStyle='danger'>{errorMessage}</Alert> :
              formattedFilePreview()
          }
        </ProgressWrapper>
      </div>
    );
  }
}

export {FilePreview, getLanguageFromFileName};
