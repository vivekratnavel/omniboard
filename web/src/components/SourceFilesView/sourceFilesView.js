import PropTypes from 'prop-types';
import React, {Component} from 'react';
import {Alert, Button} from 'react-bootstrap';
import './sourceFilesView.scss';
import ReactList from 'react-list';
import {
  Accordion,
  AccordionItem,
  AccordionItemTitle,
  AccordionItemBody
} from 'react-accessible-accordion';
import saveAs from 'file-saver';
import backend, {setDbInfo} from '../Backend/backend';
import {ProgressWrapper} from '../Helpers/hoc';
import {concatArrayBuffers, getFileExtension, parseServerError} from '../Helpers/utils';
import {imageExtensions} from '../FilePreview/imageExtensions';
import {FilePreview} from '../FilePreview/filePreview';

const MIN_ACCORDION_HEIGHT = 51;

// File preview limit set to 5MB in bytes
const FILE_PREVIEW_LIMIT = 5242880;

class SourceFilesView extends Component {
  static propTypes = {
    files: PropTypes.array.isRequired,
    type: PropTypes.string.isRequired,
    dbInfo: PropTypes.shape({
      path: PropTypes.string,
      key: PropTypes.string
    }).isRequired,
    runId: PropTypes.number.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      sourceFiles: {},
      isLoadingSourceFiles: false,
      error: '',
      isZipInProgress: false,
      isAccordionDataLoading: false,
      accordionError: ''
    };
  }

  _fetchSourceFiles = () => {
    const {runId, type} = this.props;
    if (runId && ['source_files', 'artifacts'].includes(type)) {
      this.setState({
        isLoadingSourceFiles: true,
        error: ''
      });
      let url = '';
      if (type === 'source_files') {
        url = `api/v1/files/source/${runId}`;
      } else if (type === 'artifacts') {
        url = `api/v1/files/artifacts/${runId}`;
      }

      backend.get(url).then(response => {
        const sourceFiles = response.data.reduce((files, file) => {
          files[file._id] = {
            id: file._id,
            uploadDate: file.uploadDate,
            filepath: file.filename,
            fileLength: file.length,
            chunkSize: file.chunkSize
          };
          return files;
        }, {});
        this.setState({
          sourceFiles,
          isLoadingSourceFiles: false,
          error: ''
        });
      }).catch(error => {
        const message = parseServerError(error);
        this.setState({
          isLoadingSourceFiles: false,
          error: message
        });
      });
    } else {
      this.setState({
        error: 'Invalid run id or type.'
      });
    }
  };

  _downloadFile = (fileId, fileName) => {
    return _event => {
      backend({
        url: `api/v1/files/download/${fileId}/${fileName}`,
        method: 'GET',
        responseType: 'blob'
      }).then(response => {
        saveAs(new Blob([response.data]), fileName);
      }).catch(error => {
        const message = parseServerError(error);
        this.setState({
          error: message
        });
      });
    };
  };

  _downloadAllFiles = _event => {
    const {sourceFiles} = this.state;
    const {type, runId} = this.props;
    if (Object.keys(sourceFiles).length > 0) {
      this.setState({
        isZipInProgress: true
      });
      backend({
        url: `api/v1/files/downloadAll/${runId}/${type}`,
        method: 'GET',
        responseType: 'blob'
      }).then(response => {
        this.setState({
          isZipInProgress: false
        });
        saveAs(new Blob([response.data]), `${type}-${runId}.zip`);
      }).catch(error => {
        const message = parseServerError(error);
        this.setState({
          isZipInProgress: false,
          error: message
        });
      });
    } else {
      this.setState({
        error: 'Error: No files are available to download'
      });
    }
  };

  componentDidMount() {
    setDbInfo(backend, this.props.dbInfo);
    this._fetchSourceFiles();
  }

  _handleAccordionItemChange = fileId => {
    const {sourceFiles} = this.state;
    const {files} = this.props;
    const fileInfo = files.find(file => file.file_id === fileId);
    const fileName = fileInfo.name || '';
    this.setState({
      accordionError: ''
    });
    if (sourceFiles[fileId] && !sourceFiles[fileId].data && sourceFiles[fileId].fileLength <= FILE_PREVIEW_LIMIT &&
        !imageExtensions.includes(getFileExtension(fileName))) {
      // Fetch contents of artifact/source file
      this.setState({
        isAccordionDataLoading: true
      });

      fetch(`${backend.defaults.baseURL}/api/v1/files/preview/${fileId}`)
        .then(response => {
          // Response body is a ReadableStream
          const reader = response.body.getReader();
          // Result is concatenated Array Buffer
          let result = null;
          const onStreamDone = () => {
            // Convert array buffer to string
            const resultText = new TextDecoder().decode(result);
            const updatedFileInfo = {...sourceFiles[fileId], data: resultText};
            this.setState(prevState => {
              return {
                sourceFiles: {...prevState.sourceFiles, [fileId]: updatedFileInfo},
                isAccordionDataLoading: false
              };
            });
          };

          reader.read().then(function processText({done, value}) {
            // Result objects contain two properties:
            // done  - true if the stream has already given you all its data.
            // value - some data. Always undefined when done is true.
            if (done) {
              onStreamDone();
              return;
            }

            // Concatenate array buffer
            result = concatArrayBuffers(result, value);

            // Read some more, and call this function again
            return reader.read().then(processText);
          });
        }).catch(error => {
          this.setState({
            isAccordionDataLoading: false,
            accordionError: parseServerError(error)
          });
        });
    }
  };

  _itemSizeEstimator = (_index, _cache) => {
    return MIN_ACCORDION_HEIGHT;
  };

  _renderAccordionItem = index => {
    const {files} = this.props;
    const {sourceFiles, accordionError, isAccordionDataLoading} = this.state;
    const file = files[index];
    return (
      <AccordionItem key={file.file_id} uuid={file.file_id} test-attr={'acc-item-' + index}>
        <AccordionItemTitle className='accordion__title accordion__title--animated'>
          <h5 className='u-position-relative'>
            {file.name}
            <div className='accordion__arrow' role='presentation'/>
          </h5>
          <div>{sourceFiles[file.file_id] && sourceFiles[file.file_id].filepath}</div>
        </AccordionItemTitle>
        <AccordionItemBody>
          <div className='clearfix'>
            <div className='pull-left upload-date'>Upload Date: {sourceFiles[file.file_id] && sourceFiles[file.file_id].uploadDate}</div>
            <div className='pull-right'>
              <Button test-attr={'down-btn-' + file.name} bsStyle='default' bsSize='xsmall' onClick={this._downloadFile(file.file_id, file.name)}>
                <i className='glyphicon glyphicon-download-alt'/> Download
              </Button>
            </div>
          </div>
          <FilePreview fileName={file.name} fileId={file.file_id} sourceFiles={sourceFiles}
            isLoading={isAccordionDataLoading} errorMessage={accordionError} dbInfo={this.props.dbInfo}/>
        </AccordionItemBody>
      </AccordionItem>
    );
  };

  render() {
    const {files, type} = this.props;
    const {isLoadingSourceFiles, isZipInProgress, error} = this.state;

    const errorAlert = error ? <Alert bsStyle='danger' test-attr='error-alert'>{error}</Alert> : '';
    const warningText = `Oops! There are no ${type} available for this run.`;
    const accordions = (
      <ProgressWrapper loading={isLoadingSourceFiles}>
        <div>
          <div className='download-all-wrapper clearfix'>
            <div className='pull-right'>
              <Button test-attr='down-all-btn' bsStyle='info' bsSize='small' disabled={isZipInProgress} onClick={this._downloadAllFiles}>
                {isZipInProgress ? <i className='glyphicon glyphicon-refresh glyphicon-refresh-animate'/> : <i className='glyphicon glyphicon-download-alt'/>}
              &nbsp;
                {isZipInProgress ? 'Archiving...' : 'Download All'}
              </Button>
            </div>
          </div>
          <Accordion accordion onChange={this._handleAccordionItemChange}>
            <ReactList
              itemRenderer={this._renderAccordionItem}
              length={files.length}
              type='variable'
              minSize={20}
              pageSize={10}
              threshold={300}
              itemSizeEstimator={this._itemSizeEstimator}
            />
          </Accordion>
        </div>
      </ProgressWrapper>
    );
    const content = files.length > 0 ? accordions : <Alert test-attr='warn-alert' bsStyle='warning'>{warningText}</Alert>;
    return (
      <div id='source-files-container'>
        {errorAlert}
        {content}
      </div>
    );
  }
}

export {SourceFilesView, FILE_PREVIEW_LIMIT};
