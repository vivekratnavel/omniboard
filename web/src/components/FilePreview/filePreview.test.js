import React from 'react';
import {FILE_PREVIEW_LIMIT} from '../SourceFilesView/sourceFilesView';
import {FilePreview, getLanguageFromFileName} from './filePreview';

describe('FilePreview', () => {
  let wrapper = null;
  const sourceFiles = {
    1: {
      data: 'Hello World!'
    },
    3: {
      data: 'svg_image'
    },
    4: {
      data: 'png_image'
    },
    5: {
      data: '',
      fileLength: FILE_PREVIEW_LIMIT + 1
    },
    6: {
      data: ''
    }
  };

  const dbInfo = {key: 'test', name: 'sacred', path: '/db1'};

  beforeEach(() => {
    wrapper = shallow(
      <FilePreview fileId='1' fileName='hello_world.py' sourceFiles={sourceFiles} isLoading={false} dbInfo={dbInfo}/>
    );
  });

  describe('should render correctly', () => {
    it('text files', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('svg files', () => {
      wrapper = shallow(
        <FilePreview fileId='3' fileName='test.svg' sourceFiles={sourceFiles} isLoading={false} dbInfo={dbInfo}/>
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('image files', () => {
      wrapper = shallow(
        <FilePreview fileId='4' fileName='output.png' sourceFiles={sourceFiles} isLoading={false} dbInfo={dbInfo}/>
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('when file size is too large', () => {
      wrapper = shallow(
        <FilePreview fileId='5' fileName='out.txt' sourceFiles={sourceFiles} isLoading={false} dbInfo={dbInfo}/>
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('when there is error', () => {
      wrapper = shallow(
        <FilePreview fileId='6' fileName='out.txt' sourceFiles={sourceFiles} errorMessage='error' isLoading={false} dbInfo={dbInfo}/>
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('html files', () => {
      wrapper = shallow(
        <FilePreview fileId='7' fileName='output.html' sourceFiles={sourceFiles} isLoading={false} dbInfo={dbInfo}/>
      );

      expect(wrapper).toMatchSnapshot();
    });
  });

  it('should map language from filename correctly', () => {
    let fileName = 'abc.py';

    expect(getLanguageFromFileName(fileName)).toEqual('python');
    fileName = 'unknown.abc';

    expect(getLanguageFromFileName(fileName)).toEqual('');
  });
});
