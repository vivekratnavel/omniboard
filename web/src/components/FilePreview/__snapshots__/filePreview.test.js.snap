// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilePreview should render correctly html files 1`] = `
<div>
  <ProgressWrapper
    loading={false}
  >
    <Iframe
      height="1000px"
      url="/db1/api/v1/files/preview/7/output.html"
      width="100%"
    />
  </ProgressWrapper>
</div>
`;

exports[`FilePreview should render correctly image files 1`] = `
<div>
  <ProgressWrapper
    loading={false}
  >
    <img
      alt="image"
      src="/db1/api/v1/files/download/4/output.png"
    />
  </ProgressWrapper>
</div>
`;

exports[`FilePreview should render correctly svg files 1`] = `
<div>
  <ProgressWrapper
    loading={false}
  >
    <img
      alt="image"
      src="/db1/api/v1/files/download/3/test.svg"
    />
  </ProgressWrapper>
</div>
`;

exports[`FilePreview should render correctly text files 1`] = `
<div>
  <ProgressWrapper
    loading={false}
  >
    <SyntaxHighlighter
      language="python"
      style={
        Object {
          ":not(pre) > code[class*=\\"language-\\"]": Object {
            "background": "#2d2d2d",
            "borderRadius": ".3em",
            "padding": ".1em",
            "whiteSpace": "normal",
          },
          "atrule": Object {
            "color": "#cc99cd",
          },
          "attr-name": Object {
            "color": "#e2777a",
          },
          "attr-value": Object {
            "color": "#7ec699",
          },
          "block-comment": Object {
            "color": "#999",
          },
          "bold": Object {
            "fontWeight": "bold",
          },
          "boolean": Object {
            "color": "#f08d49",
          },
          "builtin": Object {
            "color": "#cc99cd",
          },
          "cdata": Object {
            "color": "#999",
          },
          "char": Object {
            "color": "#7ec699",
          },
          "class-name": Object {
            "color": "#f8c555",
          },
          "code[class*=\\"language-\\"]": Object {
            "MozHyphens": "none",
            "MozTabSize": "4",
            "OTabSize": "4",
            "WebkitHyphens": "none",
            "background": "none",
            "color": "#ccc",
            "fontFamily": "Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace",
            "hyphens": "none",
            "lineHeight": "1.5",
            "msHyphens": "none",
            "tabSize": "4",
            "textAlign": "left",
            "whiteSpace": "pre",
            "wordBreak": "normal",
            "wordSpacing": "normal",
            "wordWrap": "normal",
          },
          "comment": Object {
            "color": "#999",
          },
          "constant": Object {
            "color": "#f8c555",
          },
          "deleted": Object {
            "color": "#e2777a",
          },
          "doctype": Object {
            "color": "#999",
          },
          "entity": Object {
            "color": "#67cdcc",
            "cursor": "help",
          },
          "function": Object {
            "color": "#f08d49",
          },
          "function-name": Object {
            "color": "#6196cc",
          },
          "important": Object {
            "color": "#cc99cd",
            "fontWeight": "bold",
          },
          "inserted": Object {
            "color": "green",
          },
          "italic": Object {
            "fontStyle": "italic",
          },
          "keyword": Object {
            "color": "#cc99cd",
          },
          "namespace": Object {
            "color": "#e2777a",
          },
          "number": Object {
            "color": "#f08d49",
          },
          "operator": Object {
            "color": "#67cdcc",
          },
          "pre[class*=\\"language-\\"]": Object {
            "MozHyphens": "none",
            "MozTabSize": "4",
            "OTabSize": "4",
            "WebkitHyphens": "none",
            "background": "#2d2d2d",
            "color": "#ccc",
            "fontFamily": "Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace",
            "hyphens": "none",
            "lineHeight": "1.5",
            "margin": ".5em 0",
            "msHyphens": "none",
            "overflow": "auto",
            "padding": "1em",
            "tabSize": "4",
            "textAlign": "left",
            "whiteSpace": "pre",
            "wordBreak": "normal",
            "wordSpacing": "normal",
            "wordWrap": "normal",
          },
          "prolog": Object {
            "color": "#999",
          },
          "property": Object {
            "color": "#f8c555",
          },
          "punctuation": Object {
            "color": "#ccc",
          },
          "regex": Object {
            "color": "#7ec699",
          },
          "selector": Object {
            "color": "#cc99cd",
          },
          "string": Object {
            "color": "#7ec699",
          },
          "symbol": Object {
            "color": "#f8c555",
          },
          "tag": Object {
            "color": "#e2777a",
          },
          "url": Object {
            "color": "#67cdcc",
          },
          "variable": Object {
            "color": "#7ec699",
          },
        }
      }
    >
      Hello World!
    </SyntaxHighlighter>
  </ProgressWrapper>
</div>
`;

exports[`FilePreview should render correctly when file size is too large 1`] = `
<div>
  <ProgressWrapper
    loading={false}
  >
    <Alert
      bsClass="alert"
      bsStyle="warning"
      closeLabel="Close alert"
    >
      File is too large to be previewed. Only files less than 5MB can be previewed.
    </Alert>
  </ProgressWrapper>
</div>
`;

exports[`FilePreview should render correctly when there is error 1`] = `
<div>
  <ProgressWrapper
    loading={false}
  >
    <Alert
      bsClass="alert"
      bsStyle="danger"
      closeLabel="Close alert"
    >
      error
    </Alert>
  </ProgressWrapper>
</div>
`;
